src/Gateways/AlipayHK/Support.php

```diff
@@ -32,34 +58,137 @@ public static function requestApi($data, $publicKey): Collection
         if (array_has($data, 'is_success') && $data['is_success'] === 'F') {
             throw new InvalidSignException('Alipay Error', 3, $data);
         }
-
-        if (!self::verifySign($data['response']['alipay'], $publicKey, true, $data['sign'])) {
+        /**
+         * $dataToBeVerify refering data['response']['alipay'] or data['response']['trade']
+         * $data['response']['trade'] is for app payment
+         * $data['response']['alipay'] is for others
+         */
+        $dataToBeVerify = array_get($data['response'], 'alipay') != null ? $data['response']['alipay'] : $data['response']['trade'];
+        if (!self::verifySign($dataToBeVerify, $publicKey, true, $data['sign'], $data['sign_type'])) {
             Log::warning('Alipay Sign Verify FAILED', $data);
 
             throw new InvalidSignException('Alipay Sign Verify FAILED', 3, $data);
         }
 
-        if (isset($data['response']['alipay']['result_code']) && $data['response']['alipay']['result_code'] === 'SUCCESS') {
+        if (isset($dataToBeVerify['result_code']) && $dataToBeVerify['result_code'] === 'SUCCESS') {
             return new Collection($data['response']['alipay']);
         }
 
+        if (isset($dataToBeVerify['trade_status'])) {
+            return new Collection($data['response']['trade']);
+        }
+
         throw new GatewayException(
             'Get Alipay API error, please check raw data for further information.',
             $data,
             json_encode($data)
         );
     }
 
-    public static function verifySign($data, $publicKey = null, $sync = false, $sign = null): bool
+    public static function verifySign($data, $publicKey = null, $sync = false, $sign = null, $signType = "MD5"): bool
     {
-        $toBeVerified  = md5(
-            mb_convert_encoding(
-                urldecode(http_build_query($data)) . $publicKey,
-                'gb2312',
-                'utf-8'
-            )
-        );
+        $toBeVerified = "";
+        if ($signType == "MD5") {
+            $toBeVerified = md5(
+                mb_convert_encoding(
+                    urldecode(http_build_query($data)) . $publicKey,
+                    'gb2312',
+                    'utf-8'
+                )
+            );
+            return $sign === $toBeVerified;
+        }
+
+        if ($signType == "RSA") {//TODO Verify with alipay public key
+            return true;
+        }
+
+    }
```

Comment:
same here, if you tends to combine `md5` and `rsa` together, don't keep the original code with new code (`rsa`) within the same method, if you want to use the same method that could support two encryption, extract them and do a switch here.